neqo-transport/src/connection.rs

```diff
@@ -746,11 +785,28 @@ impl Connection {
         out
     }
 
+    fn get_last_epoch(&mut self) -> u16 {
```

Comment:
When certificate validation fails, we are sending CONNECTION_CLOSE frame with a cert error. Without this change we would send it in a Initial packet and server would drop it. All in all we need this so that CONNECTION_CLOSE is sent in right packet.

Reply:
That seems like a separable piece from this.  For this specific PR, all we need to do is walk away from a bad connection (closing the port is fine).

If we want to properly signal CONNECTION_CLOSE, then we should do that, but I think that the way you do that is the frame builder just builds frames in both epochs.

Reply:
the frame builder is not different than before. It does exactly the same thing as before. We ere checking if we have keys for each epoch. The only thing that has change here is to send CONNECTION_CLOSE in the newst epoch not in the first one (initial epoch).

I will open a separate issue for this and remove this change from this PR. We can discuss if we want to do this in the new issue.

Reply:
A new issue would be good.  From what I can see, the frame builder should be adding the packet to every epoch on its on recognizance.  Right now, we track whether CONNECTION_CLOSE has been sent, but we really need something more fine-grained.  I see now that we only send CONNECTION_CLOSE in epoch 3, which is a mistake, so let's open an issue.

To add to that, we won't get loss signals for packets containing CONNECTION_CLOSE, we have to drive sending of the frame based on receipt of packets.